<?xml version="1.0" ?>
<sdf version="1.5">
  <world name="default">
    <!-- A global light source -->
    <include>
      <uri>model://sun</uri>
    </include>
    <!-- A ground plane -->
    <include>
      <uri>model://ground_plane</uri>
    </include>

    <!-- map -->
    <include>
      <uri>model://Map</uri>
    </include>

    <!-- model -->
    <model name="my_robot">
      <static>false</static>
      <link name='chassis'>
        <pose>0 0 .1 0 0 0</pose>
	<inertial>
	  <mass>0.2</mass>
	  <inertia>
            <ixx>0.00083</ixx>       <!-- for a box: ixx = 0.083 * mass * (y*y + z*z) -->
            <ixy>0.0</ixy>           <!-- for a box: ixy = 0 -->
            <ixz>0.0</ixz>           <!-- for a box: ixz = 0 -->
            <iyy>0.002822</iyy>      <!-- for a box: iyy = 0.083 * mass * (x*x + z*z) -->
            <iyz>0.0</iyz>           <!-- for a box: iyz = 0 -->
            <izz>0.00332</izz>       <!-- for a box: izz = 0.083 * mass * (x*x + y*y) -->
	  </inertia>
	</inertial>
        <collision name='collision'>
          <geometry>
            <box>
              <size>.4 .2 .1</size>
            </box>
          </geometry>
        </collision>
        <visual name='visual'>
          <geometry>
            <box>
              <size>.4 .2 .1</size>
            </box>
          </geometry>
        </visual>
        <collision name='caster_collision'>
          <pose>-0.15 0 -0.05 0 0 0</pose>
          <geometry>
            <sphere>
              <radius>.05</radius>
            </sphere>
          </geometry>
          <surface>
            <friction>
              <ode>
                <mu>0</mu>
                <mu2>0</mu2>
                <slip1>0.1</slip1>
                <slip2>0.1</slip2>
              </ode>
            </friction>
          </surface>
        </collision>
        <visual name='caster_visual'>
          <pose>-0.15 0 -0.05 0 0 0</pose>
          <geometry>
            <sphere>
              <radius>.05</radius>
            </sphere>
          </geometry>
        </visual>
      </link>
      
      <link name="left_wheel">
        <pose>0.1 0.13 0.1 0 1.5707 1.5707</pose>
	<inertial>
	  <mass>0.01</mass>
	  <inertia>
            <ixx>0.000026975</ixx>
            <iyy>0.000026975</iyy>
            <izz>0.00005</izz>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyz>0</iyz>
	  </inertia>
	</inertial>
        <collision name="collision">
          <geometry>
            <cylinder>
              <radius>.1</radius>
              <length>.05</length>
            </cylinder>
          </geometry>
        </collision>
        <visual name="visual">
          <geometry>
            <cylinder>
              <radius>.1</radius>
              <length>.05</length>
            </cylinder>
          </geometry>
        </visual>
      </link>

      <link name="right_wheel">
        <pose>0.1 -0.13 0.1 0 1.5707 1.5707</pose>
	<inertial>
	  <mass>0.01</mass>
	  <inertia>
            <ixx>0.000026975</ixx>
            <iyy>0.000026975</iyy>
            <izz>0.00005</izz>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyz>0</iyz>
	  </inertia>
	</inertial>
        <collision name="collision">
          <geometry>
            <cylinder>
              <radius>.1</radius>
              <length>.05</length>
            </cylinder>
          </geometry>
        </collision>
        <visual name="visual">
          <geometry>
            <cylinder>
              <radius>.1</radius>
              <length>.05</length>
            </cylinder>
          </geometry>
        </visual>
      </link>

      <link name='camera_link'>
        <pose>-0.2 0 0.1 0 0 3.14159</pose>
        <inertial>
          <mass>0.01</mass>
          <inertia>
            <ixx>0.000000664</ixx>
            <iyy>0.000000664</iyy>
            <izz>0.000000664</izz>
            <ixy>0</ixy>
            <ixz>0</ixz>
            <iyz>0</iyz>
          </inertia>
        </inertial>
        <collision name='collision'>
          <geometry>
            <box>
	      <size>0.02 0.02 0.02</size>
            </box>
          </geometry>
        </collision>
        <visual name='visual'>
          <geometry>
            <box>
	      <size>0.02 0.02 0.02</size>
            </box>
          </geometry>
        </visual>
        <sensor name='camera' type='depth'>
          <camera name='__default__'>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
	      <width>320</width>
	      <height>240</height>
            </image>
            <clip>
	      <near>0.1</near>
	      <far>100</far>
            </clip>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>1</visualize>
        </sensor>
      </link>

      <joint name="camara_joint" type="fixed">
	<child>camera_link</child>
	<parent>chassis</parent>
      </joint>
      
      <joint type="revolute" name="left_wheel_hinge">
        <pose>0 0 -0.03 0 0 0</pose>
        <child>left_wheel</child>
        <parent>chassis</parent>
        <axis>
          <xyz>0 0 1</xyz>
        </axis>
      </joint>

      <joint type="revolute" name="right_wheel_hinge">
        <pose>0 0 0.03 0 0 0</pose>
        <child>right_wheel</child>
        <parent>chassis</parent>
        <axis>
          <xyz>0 0 1</xyz>
        </axis>
      </joint>
      <plugin name='control' filename='libmy_robot_plugin.so'>
	<velocity>5</velocity>
      </plugin>
    </model>
  </world>
</sdf>
